core/src/col_metadata.rs

```diff
@@ -172,31 +176,53 @@ impl SqlGetSchemaResponse {
             self.schema.json_schema.clone().try_into()?;
         let result_set_object_schema = result_set_schema.assert_object_schema()?;
 
-        result_set_object_schema
-            .clone()
-            // 1. Access result_set_schema.properties and sort alphabetically.
-            //    This means we are sorting by datasource name.
-            .properties
-            .into_iter()
-            .sorted_by(|a, b| Ord::cmp(&a.0, &b.0))
-            // 2. map each datasource_schema to a Result of an Iterator over MongoColMetadata.
-            .map(|(datasource_name, datasource_schema)| {
-                Ok::<std::vec::IntoIter<MongoColMetadata>, Error>(
-                    Self::schema_to_col_metadata(
+        // create a map from the naming convention used by select order ([datasource name, column name]),
+        // to the schema
+        let mut processed_result_set_metadata: HashMap<Vec<String>, MongoColMetadata> =
+            result_set_object_schema
+                .clone()
+                // 1. Access result_set_schema.properties and turn into an iterator
+                .properties
+                .into_iter()
+                // 2. for each datasource, convert the schema to column metadata. Then,
+                //    turn the resulting vector of metadata into key-value pairs for the
+                //    metadata map we are creating.
+                .map(|(datasource_name, datasource_schema)| {
+                    let schema = Self::schema_to_col_metadata(
                         &datasource_schema,
                         current_db,
                         &datasource_name,
                         type_mode,
-                    )?
-                    .into_iter(),
-                )
-            })
-            // 3. flatten each Ok(inner_iterator) into the top iterator, will short circuit if an
-            //    Err is hit and just return the first Error.
-            .flatten_ok()
-            // 4. collect the Iterator<Item=Result<MongoColMetadata>> into a
-            //    Result<Vec<MongoColMetadata>>
-            .collect()
+                    )?;
+                    Ok(schema
+                        .into_iter()
+                        .map(|col| (vec![col.table_name.clone(), col.col_name.clone()], col)))
+                })
+                // flatten the key-value pairs representing the metadata into a single vector,
+                // then finally convert to a BTree
+                .flatten_ok()
+                .collect::<Result<HashMap<Vec<String>, MongoColMetadata>>>()?;
+
+        Ok(match self.select_order {
+            // in the select list order is None, for example if using an older adf version, sort
+            None => processed_result_set_metadata
```

Comment:
in theory, this shouldn't be hit when mongosql + ADL changes have been released. It was a bit more natural in JDBC, which was a nice side effect. If its considered overkill we could just error if select order is empty

Reply:
I actually like leaving this